azurerm/resource_arm_hdinsight.go

```diff
@@ -0,0 +1,1151 @@
+	hdInsightClient := meta.(*ArmClient).hdInsightClustersClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return errwrap.Wrapf("Error Parsing Azure Resource ID {{err}}", err)
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["clusters"]
+
+	future, err := hdInsightClient.Delete(ctx, resGroup, name)
+	if err != nil {
+		return fmt.Errorf("Error deleting HDInsight cluster %q (Resource Group %q): %+v", name, resGroup, err)
+	}
+
+	err = future.WaitForCompletion(ctx, hdInsightClient.Client)
+	if err != nil {
+		return fmt.Errorf("Error waiting for the deleting HDInsight cluster %q (Resource Group %q): %+v", name, resGroup, err)
+	}
+
+	return nil
+}
+
+func resourceArmHDInsightClusterUpdate(d *schema.ResourceData, meta interface{}) error {
+	hdInsightClient := meta.(*ArmClient).hdInsightClustersClient
+	ctx := meta.(*ArmClient).StopContext
+	log.Printf("[INFO] preparing arguments for AzureRM HDInsight update.")
+
+	resourceGroup := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+	tags := d.Get("tags").(map[string]interface{})
+
+	parameters := hdinsight.ClusterPatchParameters{
+		Tags: expandTags(tags),
+	}
+
+	_, err := hdInsightClient.Update(ctx, resourceGroup, name, parameters)
+	if err != nil {
+		return err
+	}
+
+	read, err := hdInsightClient.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return err
+	}
+
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read HDInsight cluster %q (resource group %q) ID", name, resourceGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmContainerRegistryRead(d, meta)
+}
+
+func expandList(interfaceList []interface{}) (map[string]interface{}, error) {
+	listLength := len(interfaceList)
+	if listLength != 1 {
+		return nil, fmt.Errorf("All lists should contain 1 item")
+	}
+	only := interfaceList[0]
+	flatmap, ok := only.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("Cannot cast list to flatmap")
+	}
+	return flatmap, nil
+}
+
+func expandAzureRmHDInsightClusterDefinition(d *schema.ResourceData) *hdinsight.ClusterDefinition {
+	clusterDefinitionInterfaceList := d.Get("cluster_definition").([]interface{})
+	clusterDefinition := &hdinsight.ClusterDefinition{}
+
+	clusterDefinitionFlatMap, err := expandList(clusterDefinitionInterfaceList)
+	if err != nil {
+		return nil
+	}
+
+	if v := clusterDefinitionFlatMap["blueprint"].(string); v != "" {
+		clusterDefinition.Blueprint = &v
+	}
+	if v := clusterDefinitionFlatMap["kind"].(string); v != "" {
+		clusterDefinition.Kind = &v
+	}
+
+	if clusterDefinitionFlatMap["component_version"] != nil {
+		componentVersion := make(map[string]*string)
+		if componentVersionFlatMap := clusterDefinitionFlatMap["component_version"].(map[string]interface{}); len(componentVersionFlatMap) != 0 {
+			for k, i := range componentVersionFlatMap {
+				if v := i.(string); v != "" {
+					componentVersion[k] = &v
```

Comment:
I think we could run into issues due to the matrix of different components and versions used in different configurations: https://docs.microsoft.com/en-us/azure/hdinsight/hdinsight-component-versioning - it could become very verbose and require tracking HDInsight version updates - thoughts?

Reply:
This is computed by the API depending on the HDInsights cluster version so I don't actually think this is required.